### PR TITLE
WIP: Make the GET DIAGNOSTICS counter 64 bit safe

### PR DESCRIPTION
**Pull request by andreasscherbaum**
Along the way update all places which work with the
number of tuples, one way or another.

Cherry-pick c970292a94e0fff468d500db430d751b6221a0b4 from upstream

### CODE REVIEW

**Comment by andreasscherbaum** [comment]
This is the successor of https://github.com/greenplum-db/gpdb/pull/2615 - I screwed up that PR beyond repair.

File: src/include/gppc.h [hunk 1]
```diff
@@ -122,8 +122,8 @@ typedef struct GppcTm
 typedef struct GppcSPIResultData
 {
 	struct GppcSPITupleTableData   *tuptable; /**< opaque pointer to the internal data */
-	uint32_t						processed; /**< number of affected rows by the operation */
-	uint32_t						current; /**< position of tuples now looking at */
+	uint64_t						processed; /**< number of affected rows by the operation */
```
Comment (hlinnaka) [inline src/include/gppc.h]: I believe the purpose of the `gppc` stuff is to keep binary compatibility of extensions across GPDB versions. I'm not sure what the role of this particular struct is, but in general, changing anything in the gppc.h header is not cool.

**Comment by hlinnaka** [comment]
Given that we already have the additional SPI_processed64 variable in GPDB, what does this buy us?

**Comment by hlinnaka** [comment]
Me & Andreas had a chat on this over IM. To summarize:

So, as a way forward, I'd suggest splitting that PR into multiple PRs, or multiple commits within the PR:
1. Change the GPDB-specific code, where applicable, to use int64 instead of int, for counting rows. I think the "reject count" stuff in COPY at least we could and should do right now, regardless of what we do with the rest of th patch. Probably there was some more GPDB-specific stuff too that makes sense regardless.
2. Cherry-pick the removal of the IncrProcessed() etc. as a separate commit. That will happen in PostgreSQL 9.0 anyway. It might cause a few merge conflicts through 8.4 and 9.0, but nothing major, and will be straightforward to resolve.
3. Let's see what's left of this patch after that.

File: contrib/spi/refint.c [hunk 2]
```diff
@@ -578,7 +578,7 @@ check_foreign_key(PG_FUNCTION_ARGS)
 
 		snprintf(ident, sizeof(ident), "%s$%u", trigger->tgname, rel->rd_id);
 		plan = find_plan(ident, &FPlans, &nFPlans);
-		ret = SPI_execp(plan->splan[r], kvals, NULL, tcount);
+		ret = SPI_execp(plan->splan[r], kvals, NULL, (int64) tcount);
```
Comment (andreasscherbaum) [inline contrib/spi/refint.c]: Removed

File: src/backend/access/external/fileam.c [hunk 3]
```diff
@@ -355,7 +355,7 @@ external_endscan(FileScanDesc scan)
 	if (scan->fs_pstate != NULL && scan->fs_pstate->errMode != ALL_OR_NOTHING)
 	{
 		if (Gp_role == GP_ROLE_EXECUTE)
-			SendNumRowsRejected(scan->fs_pstate->cdbsreh->rejectcount);
+			SendNumRowsRejected((int64) scan->fs_pstate->cdbsreh->rejectcount);
```
Comment (andreasscherbaum) [inline src/backend/access/external/fileam.c]: Removed.

File: src/backend/commands/copy.c [hunk 4]
```diff
@@ -212,7 +212,7 @@ else \
 												   sizeof(char) + 1 + 24); \
 \
```
Comment (andreasscherbaum) [inline src/backend/commands/copy.c]: Expanded the string.

File: src/include/gppc.h (hunk 1 above)
Comment (andreasscherbaum) [inline src/include/gppc.h]: Well, it stores the result of SPI_processed(), so one way or another this code must be touched. But this can be a separate PR, after I split this one up.

**Comment by ashwinstar** [comment]
@andreasscherbaum Is anything happening on this OR or things moved to https://github.com/greenplum-db/gpdb/pull/2679 and this can be closed out.

**Comment by andreasscherbaum** [comment]
@ashwinstar This PR is the overall patch, including all proposed changes.
#2679 is an extract which just includes SPI and pl/pgSQL changes.

I still need this one here.

**Comment by dyozie** [comment]
@andreasscherbaum - when this PR gets merged does it mean that GET DIAGNOSTICS will be supported in GPDB, or is there other work to be done?  I noticed that postgresql doesn't support GET DIAGNOSTICS even through version 9.2 (https://www.postgresql.org/docs/9.2/static/unsupported-features-sql-standard.html).

**Comment by andreasscherbaum** [comment]
@dyozie This is about "GET DIAGNOSTICS" in pl/pgSQL, and this is supported: https://www.postgresql.org/docs/8.0/static/plpgsql-statements.html

**Comment by xinzweb** [comment]
@andreasscherbaum do we still need to keep this PR around? @Eulerizeit is that something already on your radar? To remove it from the PR reminder, just add 'ToDo' tag. Thanks.

**Comment by andreasscherbaum** [comment]
@xinzweb Still need it. There is a number of split-off patches based on this one, but they are all blocked on #4669 right now.

**Comment by xinzweb** [comment]
Thanks a lot. I exclude all the WIP from the Pull Reminder. :) Thanks a lot for the update.

**Comment by pivotal-mike** [comment]
@ascherbaum-pivotal Are we good to close or merge this given the 64-bit patches are in?
